source/slang/slang-check-decl.cpp

```diff
@@ -7981,19 +8326,325 @@ namespace Slang
+                    SLANG_ASSERT(memberIndex < members.getCount());
+                    member = members[memberIndex++].getDecl();
+                }
+
+                // We need to ensure member is `no_diff` if it cannot be differentiated, `ctor` modifiers do not matter
+                // in this case since member-wise ctor is always differentiable or "treat as differentiable".
+                if (!isTypeDifferentiable(member->getType()) || member->hasModifier<NoDiffModifier>())
+                {
+                    auto noDiffMod = m_astBuilder->create<NoDiffModifier>();
+                    noDiffMod->loc = param->loc;
+                    addModifier(param, noDiffMod);
+                }
+
+                addCudaHostModifierIfRequired(ctor, member, foundCudaHostModifier);
+
+                MemberExpr* memberExpr = m_astBuilder->create<MemberExpr>();
+                memberExpr->baseExpression = thisExpr;
+                memberExpr->declRef = member->getDefaultDeclRef();
+                memberExpr->scope = ctor->ownedScope;
+                memberExpr->loc = member->loc;
+                memberExpr->name = member->getName();
+                memberExpr->type = DeclRefType::create(getASTBuilder(), memberExpr->declRef);
+
+                auto assign = m_astBuilder->create<AssignExpr>();
+                assign->left = memberExpr;
+                assign->right = paramExpr;
+                assign->loc = paramExpr->loc;
+
+                auto stmt = m_astBuilder->create<ExpressionStmt>();
+                stmt->expression = assign;
+                stmt->loc = assign->loc;
+
+                seqStmtChild->stmts.add(stmt);
+            }
+            seqStmt->stmts.insert(ctorInfo.m_insertOffset++, seqStmtChild);
         }
 
-        if (structDeclInfo.defaultCtor)
+        // Compiler generated ctor may be destroyed
+        bool destroyedDefaultCtor = false; 
+        if(structDeclInfo.m_defaultCtor
+            && structDeclInfo.m_defaultCtor->containsOption(ConstructorTags::Synthesized))
         {
-            auto seqStmt = as<SeqStmt>(as<BlockStmt>(structDeclInfo.defaultCtor->body)->body);
-            if (seqStmt && seqStmt->stmts.getCount() == 0)
+            if (!structDeclInfo.m_defaultCtor->body)
+                destroyedDefaultCtor = true;
+            else if (auto block = as<BlockStmt>(structDeclInfo.m_defaultCtor->body))
+            {
+                if (as<SeqStmt>(block->body)->stmts.getCount() == 0)
+                    destroyedDefaultCtor = true;
+            }
+            if (destroyedDefaultCtor)
             {
-                structDecl->members.remove(structDeclInfo.defaultCtor);
+                structDecl->members.remove(structDeclInfo.m_defaultCtor);
                 structDecl->invalidateMemberDictionary();
                 structDecl->buildMemberDictionary();
             }
         }
+
+        // Fill in our $ZeroInit
+        if (auto zeroInitListFunc = findZeroInitListFunc(structDecl))
```

Comment:
I think we shouldn't need this at all because our IR already knows how to zerzo init a thing.
We can simply do:
```
this = DefaultConstruct(This);
```
at the begining of the ctor, where `DefaultConstruct(T)` lowers to `kIROp_DefaultConstruct(T)`

Reply:
We could add a flag to DefaultConstructExpr to remove the need for $ZeroInit.

To elaborate, DefaultConstructExpr does not handle initExpr on a struct member (from what I remember).

We need a flag that we can set to tell Slang when lowering a DefaultConstructExpr that it should take into account initExpr of varDecl's.

Reply:
This change already did the 
`this = DefaultConstruct(This);`

but it looks like `DefaultConstruct` will make everything to 0, and doesn't handle the initExpr and the base class.

Reply:
Close this conversion for now, please re-open it if we want to re-visit it.